src/openms/include/OpenMS/ANALYSIS/ID/IDBoostGraph.h

```diff
@@ -0,0 +1,278 @@
+// $Authors: Julianus Pfeuffer $
+// --------------------------------------------------------------------------
+
+#ifndef OPENMS_ANALYSIS_ID_IDBOOSTGRAPH_H
+#define OPENMS_ANALYSIS_ID_IDBOOSTGRAPH_H
+
+#include <OpenMS/CONCEPT/LogStream.h>
+#include <OpenMS/CONCEPT/Types.h>
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/METADATA/ProteinIdentification.h>
+#include <OpenMS/METADATA/PeptideIdentification.h>
+
+#include <vector>
+
+#include <boost/function.hpp>
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/filtered_graph.hpp>
+#include <boost/graph/properties.hpp>
+#include <boost/variant.hpp>
+#include <boost/variant/detail/hash_variant.hpp>
+#include <boost/variant/static_visitor.hpp>
+
+namespace OpenMS
+{
+
+  /**
+   @brief Creates and maintains a boost graph based on the OpenMS ID datastructures
+
+   For finding connected components and applying functions to them.
+   Currently assumes that all PeptideIdentifications are from the ProteinID run that is given.
+   Please make sure this is right.
+   VERY IMPORTANT NOTE: If you add Visitors here, make sure they do not touch members of the
+   underlying ID objects that are responsible for the graph structure. E.g. the (protein/peptide)_hits vectors
+   or the lists in ProteinGroups. You can set information like scores or metavalues, though.
+
+   @ingroup Analysis_ID
+   */
+  class IDBoostGraph
+  {
+
+  public:
+
+    BOOST_STRONG_TYPEDEF(char, PeptideCluster)
+    BOOST_STRONG_TYPEDEF(char, ProteinGroup)
+
+    //typedefs
+    //typedef ProteinIdentification::ProteinGroup ProteinGroup;
+
+    typedef boost::variant<ProteinHit*, ProteinGroup*, PeptideCluster*, PeptideHit*> IDPointer;
+    typedef boost::variant<const ProteinHit*, const ProteinGroup*, const PeptideCluster*, const PeptideHit*> IDPointerConst;
+    //TODO check the impact of different datastructures to store nodes/edges (maybe also use directed graph?)
+    typedef boost::adjacency_list <boost::setS, boost::vecS, boost::undirectedS, IDPointer> Graph;
+    typedef boost::adjacency_list <boost::setS, boost::vecS, boost::undirectedS, IDPointerConst> GraphConst;
+    typedef boost::graph_traits<Graph>::vertex_descriptor vertex_t;
+    typedef boost::graph_traits<Graph>::edge_descriptor edge_t;
+    typedef boost::filtered_graph<Graph, boost::function<bool(edge_t)>, boost::function<bool(vertex_t)> > FilteredGraph;
+    typedef std::set<IDBoostGraph::vertex_t> ProteinNodeSet;
+    typedef std::set<IDBoostGraph::vertex_t> PeptideNodeSet;
+
+    /// Constructor
+    IDBoostGraph(ProteinIdentification &proteins, std::vector<PeptideIdentification>& idedSpectra);
+
+    /// Do sth on connected components
+    void applyFunctorOnCCs(std::function<void(FilteredGraph &)> functor);
+
+    /// Add intermediate nodes to the graph that represent indist. protein groups and peptides with the same parents
+    /// this will save computation time and oscillations later on.
+    void clusterIndistProteinsAndPeptides();
```

Comment:
is it mandatory to call this method? or just to speed up the computation?

Reply:
Yes this is not used yet. It's for later and yes mostly for saving duplicate work.